src/main/java/org/alliancegenome/curation_api/services/GeneExpressionAnnotationService.java

```diff
@@ -0,0 +1,135 @@
+
+	@Inject VocabularyTermService vocabularyTermService;
+
+	@Inject	PersonDAO personDAO;
+
+	@Inject PersonService personService;
+
+	@Override
+	protected void init() {
+		setSQLDao(geneExpressionAnnotationDAO);
+	}
+
+	@Override
+	public GeneExpressionAnnotation deprecateOrDeleteAnnotationAndNotes(Long id, Boolean throwApiError, String loadDescription, Boolean deprecate) {
+		GeneExpressionAnnotation geneExpressionAnnotation = geneExpressionAnnotationDAO.find(id);
+
+		if (geneExpressionAnnotation == null) {
+			String errorMessage = "Could not find Gene Expression Annotation with id: " + id;
+			if (throwApiError) {
+				ObjectResponse<GeneExpressionAnnotation> response = new ObjectResponse<>();
+				response.addErrorMessage("id", errorMessage);
+				throw new ApiErrorException(response);
+			}
+			log.error(errorMessage);
+			return null;
+		}
+		if (deprecate) {
+			if (!geneExpressionAnnotation.getObsolete()) {
+				geneExpressionAnnotation.setObsolete(true);
+				if (authenticatedPerson.getId() != null) {
+					geneExpressionAnnotation.setUpdatedBy(personDAO.find(authenticatedPerson.getId()));
+				} else {
+					geneExpressionAnnotation.setUpdatedBy(personService.fetchByUniqueIdOrCreate(loadDescription));
+				}
+				geneExpressionAnnotation.setDateUpdated(OffsetDateTime.now());
+				return geneExpressionAnnotationDAO.persist(geneExpressionAnnotation);
+			} else {
+				return geneExpressionAnnotation;
+			}
+		} else {
+			geneExpressionAnnotationDAO.remove(id);
+		}
+		return null;
+	}
+
+	public List<Long> getAnnotationIdsByDataProvider(BackendBulkDataProvider dataProvider) {
+		Map<String, Object> params = new HashMap<>();
+		params.put(EntityFieldConstants.DATA_PROVIDER, dataProvider.sourceOrganization);
+		if (StringUtils.equals(dataProvider.sourceOrganization, "RGD") || StringUtils.equals(dataProvider.sourceOrganization, "XB")) {
+			params.put(EntityFieldConstants.PA_SUBJECT_TAXON, dataProvider.canonicalTaxonCurie);
+		}
+		List<Long> annotationIds = geneExpressionAnnotationDAO.findIdsByParams(params);
+		return annotationIds;
+	}
+
+	@Transactional
+	@Override
+	public GeneExpressionAnnotation upsert(GeneExpressionFmsDTO geneExpressionFmsDTO, BackendBulkDataProvider dataProvider) throws ObjectUpdateException {
+		if (geneExpressionAnnotationFmsDTOValidator.isExpressionAnnotationValid(geneExpressionFmsDTO)) {
+			GeneExpressionAnnotation geneExpressionAnnotation = new GeneExpressionAnnotation();
+			String uniqueId = geneExpressionAnnotationUniqueIdHelper.generateUniqueId(geneExpressionFmsDTO);
+			SearchResponse<GeneExpressionAnnotation> annotationDB = geneExpressionAnnotationDAO.findByField("uniqueId", uniqueId);
+			if (annotationDB != null && annotationDB.getSingleResult() != null) {
+				geneExpressionAnnotation = annotationDB.getSingleResult();
+			}
+			geneExpressionAnnotation.setUniqueId(uniqueId);
+			Gene expressionAnnotationSubject = geneService.findByIdentifierString(geneExpressionFmsDTO.getGeneId());
```

Comment:
I don't really understand why you've put all this sort of code in the service class rather than following the paradigm we've used elsewhere of setting fields in the validation classes.  You're doing a lookup against the database here to find the gene based on the curie, but you've already done exactly the same lookup in the validation class.

Reply:
Yes, there are 2 calls. Initially I ran the MOD files  validating all the fields to find data problems and that code structure stuck. I will change it